src/v/lsm/io/cloud_persistence.cc

```diff
@@ -0,0 +1,471 @@
+        auto file = ss::open_file_dma(
+          filepath.native(),
+          ss::open_flags::create | ss::open_flags::rw
+            | ss::open_flags::truncate);
+        auto output_stream
+          = co_await ss::with_file_close_on_failure(
+              std::move(file),
+              [content_length](ss::file& f) {
+                  return f.allocate(0, content_length)
+                    .then([f = std::move(f)] mutable {
+                        return ss::make_file_output_stream(
+                          std::move(f), ss::file_output_stream_options{});
+                    });
+              })
+              .then_wrapped(
+                [&input_stream](ss::future<ss::output_stream<char>> fut) {
+                    if (fut.failed()) {
+                        // Ensure that we close the input stream.
+                        return fut.finally(
+                          [&input_stream] { return input_stream.close(); });
+                    }
+                    return fut;
+                });
+        co_await ss::copy(input_stream, output_stream)
+          .finally([&output_stream] { return output_stream.close(); })
+          .finally([&input_stream] { return input_stream.close(); });
+        co_return content_length;
+    }
+
+    std::filesystem::path staging_path(std::string_view name) {
+        return _staging / name;
+    }
+
+    cloud_storage_clients::object_key cloud_key(std::string_view name) {
+        return join(_prefix, name);
+    }
+
+    std::filesystem::path _staging;
+    cloud_io::remote* _remote;
+    cloud_storage_clients::bucket_name _bucket;
+    cloud_storage_clients::object_key _prefix;
+    ss::abort_source _as;
+    ss::gate _gate;
+};
+
+class cloud_metadata_persistence : public metadata_persistence {
+public:
+    cloud_metadata_persistence(
+      cloud_io::remote* remote,
+      cloud_storage_clients::bucket_name bucket,
+      const cloud_storage_clients::object_key& prefix)
+      : _remote(remote)
+      , _bucket(std::move(bucket))
+      , _manifest_key(join(prefix, "MANIFEST")) {}
+
+    ss::future<std::optional<iobuf>> read_manifest() override {
+        _as.check();
+        auto _ = _gate.hold();
+        retry_chain_node root{_as};
+        iobuf b;
+        // TODO: Use CAS or term (term requires listing, which is fine)
+        auto result = co_await _remote->download_object({
+          .transfer_details = {
+            .bucket = _bucket,
+            .key = _manifest_key,
+            .parent_rtc = root,
+          },
+          .display_str = "LSM Manifest download",
+          .payload = b,
+          .expect_missing = true,
+        });
+        co_return check_result(result) ? std::make_optional(std::move(b))
+                                       : std::nullopt;
+    }
+
+    ss::future<> write_manifest(iobuf b) override {
+        _as.check();
+        auto _ = _gate.hold();
+        retry_chain_node root{_as};
+        // TODO: Use CAS or term
```

Comment:
this is a pretty big TODO :P -- is there a proposal written down? i think there are a lot of details here... but yeh can definitely be introduced later.